pennylane/_qubit_device.py

```diff
@@ -688,3 +697,96 @@ def sample(self, observable):
+
+    def adjoint_jacobian(self, tape):
+        """Implements the adjoint method outlined in
+        `Jones and Gacon <https://arxiv.org/abs/2009.02823>`__ to differentiate an input tape.
+
+        After a forward pass, the circuit is reversed by iteratively applying inverse (adjoint)
+        gates to scan backwards through the circuit. This method is similar to the reversible
+        method, but has a lower time overhead and a similar memory overhead.
+
+        .. note::
+            The adjoint differentation method has the following restrictions:
+
+            * As it requires knowledge of the statevector, only statevector simulator devices can be
+              used.
+
+            * Only expectation values are supported as measurements.
+
+        Args:
+            tape (.QuantumTape): circuit that the function takes the gradient of
+
+        Returns:
+            np.ndarray: the derivative of the tape with respect to trainable parameters.
+            Dimensions are ``(len(observables), len(trainable_params))``.
+
+        Raises:
+            QuantumFunctionError: if the input tape has measurements that are not expectation values
+                or contains a multi-parameter operation aside from :class:`~.Rot`
+        """
+
+        for m in tape.measurements:
+            if m.return_type is not qml.operation.Expectation:
+                raise qml.QuantumFunctionError(
+                    "Adjoint differentiation method does not support"
+                    f" measurement {m.return_type.value}"
+                )
+
+            if not hasattr(m.obs, "base_name"):
+                m.obs.base_name = None  # This is needed for when the observable is a tensor product
+
+        # Perform the forward pass
+        self.reset()
+
+        # Consider using caching and calling lower-level functionality. We just need the state
+        # without postprocessing https://github.com/PennyLaneAI/pennylane/pull/1032/files#r563441040
+        self.execute(tape)
+
+        phi = self._reshape(self.state, [2] * self.num_wires)
+
+        lambdas = [self._apply_operation(phi, obs) for obs in tape.observables]
+
+        expanded_ops = []
+        for op in reversed(tape.operations):
+            if op.num_params > 1:
+                if isinstance(op, qml.Rot) and not op.inverse:
+                    ops = op.decomposition(*op.parameters, wires=op.wires)
+                    expanded_ops.extend(reversed(ops))
+                else:
+                    raise QuantumFunctionError(
+                        f"The {op.name} operation is not supported using "
+                        'the "adjoint" differentiation method'
+                    )
+            else:
+                if op.name not in ("QubitStateVector", "BasisState"):
+                    expanded_ops.append(op)
+
+        jac = np.zeros((len(tape.observables), len(tape.trainable_params)))
+        dot_product_real = lambda a, b: self._real(sum(self._conj(a) * b))
```

Comment:
Is there a reason for not using `np.dot(a.conj(), b).real` instead? Actually, I'm not really sure why these NumPy methods are renamed to `_real` and `_conj` in the first place. :thinking:

Reply:
The idea is to let the device choose how to implement these linear algebra methods, in case the underlying object is not a numpy array. For example, we might have a tensor with `default.qubit.tf`. This can largely be replaced by `qml.math` in future.

We then can't do things like `a.conj()` because it's not uniformly defined across the different interfaces.

Anyway I don't think there's a strong use case for using `default.qubit.tf` - since we mainly want to do that with the backprop diff-method. But I was just being cautious here.